### ISSUE TITLE
heavy flickering when running on macOS 10.14.1

### ISSUE THREAD

**Issue by vasquesbc** [OpenDUNE/OpenDUNE#325]
i'am using release: opendune-0.9-macosx-x86_64-sdl2 on macOS 10.14.1 and experiencing heavy flickering especially if the screen fades between different menus/screens.
if there is anything i can provide to help investigating this issue, please let me know.

**Comment by miniupnp** [comment]
can you build from sources ?
You'll need to have XCode command line tools installed and set up.
And install sdl2 from [homebrew](https://brew.sh/) or [macports](https://www.macports.org/)
I hope everything builds fine on Mojave / 10.14

**Comment by vasquesbc** [comment]
i got XCode 10.1 installed & done „brew install sdl2“
however: make seems to still miss sdl?

```
[SRC] Compiling video/video_sdl2.c
/Users/merolla.c/Downloads/OpenDUNE-0.9/src/video/video_sdl2.c:5:10: fatal error: 'SDL_image.h' file not found
#include <SDL_image.h>
         ^~~~~~~~~~~~~
1 error generated.
make[1]: *** [video/video_sdl2.o] Error 1
make: *** [all] Error 1
```

**Comment by vasquesbc** [comment]
ok - got it compiled & running (i've learned, that sdl_image is seperate from sdl/sdl2).
unfortunately the flickering is still there.

**Comment by vasquesbc** [comment]
i've done a screen recording of the issue.
it seams to cohere with mouse cursor movement.
[https://youtu.be/tBSJg7p1TUc](url)

**Comment by miniupnp** [comment]
That's _probably_ a SDL2 issue...  have you tried other SDL2 programs ?

**Comment by vasquesbc** [comment]
> That's _probably_ a SDL2 issue... have you tried other SDL2 programs ?

could you drop me a link to a similar project in terms of used SDL version?

i've testet the current code on another machine with macOS 10.13.6 - same issue. it doesnt seem to be an 10.14 issue.

**Comment by miniupnp** [comment]
OK. I need to set up a mac to test...

**Comment by atalkingegg** [comment]
I'm seeing this as well. Other SDL2 projects seem fine (BZFlag works great). Using libsdl2 2.0.9_1 from macports on macOS 10.13.6, on a MacBook Pro with an Nvidia GT750M gpu.

**Comment by atalkingegg** [comment]
limiting the frame rate to 30fps changes the flicker to page tears, as if it's trying to sync to vertical refresh, but the timing is off. This is in window and full screen mode. The opening animations show two frames stuttering at a time, as if double buffering is somehow broken. Hope this helps.

**Comment by miniupnp** [comment]
that is strange the OpenDUNE SDL2 code is pretty straight forward.
For each "tick" :
```
SDL_LockTexture()
SDL_UnlockTexture()
SDL_RenderCopy()
SDL_RenderPresent()
```

**Comment by atalkingegg** [comment]
Adding a "SDL_RenderClear(s_renderer);" line after the SDL_RenderPresent,
eliminates all the old frame stuttering, but also makes it so when theres a
pause in screen updates, blank screens are presented instead. During the
game, it's still flashing, but now there's not old graphics being presented.

Replacing  SDL_CreateWindowAndRenderer with

s_window = SDL_CreateWindow(
                window_caption,
                SDL_WINDOWPOS_CENTERED,
                SDL_WINDOWPOS_CENTERED,
                SCREEN_WIDTH * s_screen_magnification,
                SCREEN_HEIGHT * s_screen_magnification,
                window_flags);
s_renderer = SDL_CreateRenderer(s_window, -1, SDL_RENDERER_ACCELERATED |
SDL_RENDERER_PRESENTVSYNC);

makes the random position screen tearing go away, but the flickering /
flashing is still there.

It looks like every call to Video_Tick, it will draw the screen even if
there's no graphics updates, which results in blank screens (if RenderClear
is used), or old frames being presented. It would be a much better use of
CPU resources to never call SDL_RenderPresent if no graphics updates are
needed, and somewhere still in there I suspect are some incorrect
assumptions about which graphics buffers are being drawn to.

I was able to play through the first level, it did crash after entering a
name at the end but it did manage to get through the level. It also crashes
on forced exits.

Really awesome project! I'm very happy it exists and that I found it, and
would be happy to help get it working properly on Mac.
Thanks!

**Comment by miniupnp** [comment]
OK there is a small bug
https://github.com/OpenDUNE/OpenDUNE/blob/master/src/video/video_sdl2.c#L669
_SDL_RenderPresent()_ is called even if _Video_DrawScreen()_ did nothing  because the screen needed no repaint. However, _SDL_RenderPresent()_ should still show the same picture...

**Comment by miniupnp** [comment]
@atalkingegg please could you tell me if https://github.com/miniupnp/OpenDUNE/commit/7c91f7b22d482259186d4002a882dfcbcde1a90b improves things ?

**Comment by miniupnp** [comment]
I just tested 7c91f7b on a HighSierra Mac, it looks like it fixes the problem

**Comment by miniupnp** [comment]
well in fact it is better, but not completely fixed :(

**Comment by atalkingegg** [comment]
Yes. Way better than it was! Looks solid now!
I found that if you set the frame rate to 60, it seams nearly perfect. Set
to 30, some of the intro titles flash in brightness.
I'll have to get dosbox/dune2 up and running and compare, I suspect there's
a few minor timing differences on fades and moving the field around. Also
there's a grungy filter sweep during the intro (Adlib FM) that I used to
really liked, it would be great to get that back.
Thanks!
